### DISCUSSION TITLE
QuickStart from add-osmt-dev-util branch build won't load/complete auth, stuck on login?

### DISCUSSION THREAD

**Discussion by sleeman1** [wgu-opensource/osmt#125] [Asking for Help] [answered]
Docker is running on an Ubuntu 18.04.3 VM

Using the add-osmt-dev-util branch and quickstart via osmt_cli.sh -q

The install using osmt_cli.sh -q does call out a whitespace error in the docker-compose.yml file - "xpack.security.enabled: false" Removing the space between : and false seems to do the trick.

I am not seeing any obvious errors as the setup runs.

Once the setup appears complete and I go to http://localhost:8080 loading http://localhost:8080/login?return=%2Fskills where the site hangs.

Okta account is set up and tests appear fine. If I go to http://localhost:8080/oauth2/authorization/okta It accepts the login and send me back to http://localhost:8080/login?return=%2Fskills  where the page just sits, and "spins" See image below.

Thoughts?

![image](https://user-images.githubusercontent.com/64618524/150563663-cf854892-97a0-440c-939e-66f01b6dfb45.png)

If I let it go long enough it will "Log me out"
![image](https://user-images.githubusercontent.com/64618524/150570997-276cad69-71d8-434c-8b40-be3ba01dc02d.png)

**Comment by JohnKallies** [comment]
Your observations are spot on... there's a typo in the docker-compose files. A patch was just merged!

Until the PR for feature/add-osmt-dev-util is merged, please make these changes locally -- https://github.com/wgu-opensource/osmt/pull/126

**Comment by JohnKallies** [comment, marked as answer]
Regarding the spinning...
- Can you please look at your network tab in your browser debugger. Is it issuing request after request?
- Is there anything helpful in ```docker logs osmt_app_1```? Anything problematic after the log that says `edu.wgu.osmt.ApplicationKt: Started ApplicationKt in xxx seconds` 

Also, since you have the osmt_cli utility, Can you please paste the output of ```./osmt_cli.sh -v```?

**Reply by sleeman1** [reply]
It does not appear to be sending multiple requests:
![image](https://user-images.githubusercontent.com/64618524/150574027-f32cdd64-5934-4f6d-a929-6b829f024b28.png)


I am seeing this error:
![image](https://user-images.githubusercontent.com/64618524/150573916-54d12b0d-ff8d-41a5-af7d-ad8895badf4d.png)

In the docker logs I see a warning but I dont think it's relatred?
Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.16/security-minimal-setup.html to enable security.&quot;],[299 Elasticsearch-7.16.2-2b937c44140b6559905130a8650c64dbd0879cfb &quot;[ignore_throttled] parameter is deprecated because frozen indices have been deprecated. Consider cold or frozen tiers in place of frozen indices.&quot;]

Here is a copy of the log file [dockerlogsUW.txt](https://github.com/wgu-opensource/osmt/files/7915644/dockerlogsUW.txt)

I am seeing an interesting error in the validation checker log : [Validation_log.txt](https://github.com/wgu-opensource/osmt/files/7915688/Validation_log.txt)  

INFO:  Checking /home/csleeman/osmt/osmt-quickstart.env for invalid default values (xxxxxx)
INFO:  No invalid default values in /home/csleeman/osmt/osmt-quickstart.env

INFO:  Checking /home/csleeman/osmt/api/osmt-dev-stack.env for invalid default values (xxxxxx)
ERROR: /home/csleeman/osmt/api/osmt-dev-stack.env should be updated with valid OAUTH2 values

I copy pasted the values across, but this says something is not correct. I used the example in the ReadMe then replaced with info from Okta.

I have a meeting at 10 PST but will look at the Okta info again after I am free around 10:30. 

If I call to the env file on startup will it update the changes/pass them in every time?

**Reply by JohnKallies** [reply]
Huh. Ok...

For today, you can ignore the 404 on whitelabel.json in Quickstart. We've got some work to do around white labeling. Search for "whitelabel" in Issues if you are interested in helping with that.

+2 on the ElasticSearch warnings in the Docker logs not being relevant. These started happening with some log4j patching we did, and will be resolved with a component upgrade. See comments in application.properties for more context.

Re: Validation_log.txt
- the error for Java not found isn't relevant in a Quickstart config
- The error reported for "xxxxxx" is in your osmt-dev-stack.env file, which is for doing active development work in the API or UI modules. The docker logs for your Quickstart containers show appropriate spring.security.oauth2.client values.

Try this... (this will require a Java 11 and Maven 3.8ish on your VM. OpenJDK works fine)
- stop the Quickstart
- run `./osmt_cli.sh -d` to start the dev stack Docker services
- run `./osmt_cli.sh -s` to start the Spring Boot app vis Maven
-- if that give you trouble, run `mvn clean install -DskipTests` from the project root, then try again

Open your browser to http://hostname:8080
-- any trouble in the browser? In the Spring Boot logs?

**Reply by sleeman1** [reply]
FYI the ./osmt_cli.sh -d will error on a whitespace error in the omst\docker\base-images\dev-stack.yml  file - "xpack.security.enabled: false"
Removing this space allows for the script to run and start the 3 expected containers. 

I configured the osmt-dev-stack.env located in \api adding the Okta info.

When I run ./osmt_cli.sh -s I get an error and the setup appears to hang for a while before spitting out more error :
INFO:  Starting OSMT via Maven Spring Boot plug-in (Maven log output suppressed)...
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[ERROR] Failed to execute goal on project osmt-api: Could not resolve dependencies for project edu.wgu.osmt:osmt-api:jar:1.0.5-SNAPSHOT: Could not find artifact edu.wgu.osmt:osmt-ui:jar:1.0.5-SNAPSHOT in spring-repo (https://repo.spring.io/release) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
ERROR: Error running OSMT Spring app with Maven. Returning to project root (/home/csleeman/osmt)


I ran mvn clean install -DskipTests (took a few minutes) but this did not change the results when running -d and -s ./osmt_cli.sh

**Reply by JohnKallies** [reply]
Ok. At present, the API module has a dependency on the UI module's jar. Did you run `mvn clean install -D skipTests` from the project root? 

`mvn install` compiles, tests, and packages the module in question, but also added the artifact to your local Maven repo (usually in ~/.m2/repository). When resolving dependencies, Maven looks there first. Because it's looking for `osmt-ui:jar:1.0.5-SNAPSHOT` in spring-repo, it didn't find it in your local.

So it looks like you ran `mvn clean install -D skipTests` from the api module. Try it from the project root instead.

**Reply by sleeman1** [reply]
OK that worked, good catch. The environment stood up fine after a clean mvn install.

I went to test with Firefox but It was still not working. I thought to myself. Hmmm could this be something stupid like popup blocker/browser related? I install chrome to test if the issue was browser related. **Chrome worked!**  I log in fine with my Okta account!

I looked at the Firefox version and it was way out of date. Firefox would not auto update so out of date. I installed 96.0 and sure enough the site loads fine! 

OK, so now I have this feeling I had the site working from the start but it was the browser not passing login/loading the site correct. I have no choice and must work backwards to the default  wgu-opensource/osmt  and try running the quick start from the .yml file. ~~Before I do that I will try a -q setup from the add-osmt-dev-util branch~~ (Works!)

Worst case we now know we can build this demo out of the add-osmt-dev-util branch using the bash script and the dev path.

There was defiantly errors that we worked through and solved along the way. However, when we would get the page to load but hang on auth, that was Firefox being the killjoy.

Feels so good to get to this point I almost forgot that this is just the demo/dev sites to set up and share with the team here at the UW. LOL

**Reply by JohnKallies** [reply]
That's great news. It's complex, and it's not like there's a handy MSI file to double click...

We are working to streamline it, and community contribution helps this immensely. The osmt_cli tool came out of watching users in this Discussion board, and it's not even merged yet. Also, We are less than a week out from pushing a bona fide OSMT app layer Docker image to DockerHub that doesn't need Okta. Quickstart then becomes a docker-compose up, even in Windows (in theory).

Good feedback on FireFox. I prefer FF in my personal life, but WGU internally nudges me toward Chrome, so I go along to get along. It would be smart for us to get broader testing and a clear understanding on supported browsers.

Once again, awesome news on getting it working! Keep in mind that the Docker volume is _the only data persistence_ in play, so point your OSMT at a less ephemeral MySQL database before putting too much time into building skills libraries and collections. Have a great weekend!

**Comment by JohnKallies** [comment]
For what it's worth, We are working to wrap up changes to Quickstart that will no longer require OAuth2/OIDC, just to make this more point-and-click.

**Comment by JohnKallies** [comment]
Chad, I missed 1 question...
> If I call to the env file on startup will it update the changes/pass them in every time?

That would be up to docker-compose. I don't exactly know, but from ignorance, I often redeploy containers that I could possibly just restart. If you find out, I'd appreciate knowing what you learn.
